main.ts

```diff
@@ -65,28 +65,36 @@ async function enableJoinRequests(username: string) {
   }
 }
 
+const timestamps = new Map<string, number[]>();
+const indexes = new Map<string, number>();
 const limit = 30;
 const timeframe = 1 * HOUR;
 function incr(username: string) {
-  const count = `${username}_count`, lastReset_ = `${username}_lastReset`;
-
-  let current = Number(localStorage.getItem(count)) || 0;
-  if (!current) {
-    localStorage.setItem(lastReset_, String(Date.now()));
-  } else {
-    const lastReset = Number(localStorage.getItem(lastReset_) || 0);
-    if (lastReset && Date.now() - lastReset >= timeframe) {
-      current = 0;
-      localStorage.removeItem(lastReset_);
+  username = username.toLowerCase();
+  const timestamps_ = (() => {
+    const v = timestamps.get(username);
+    if (v) {
+      return v;
+    } else {
+      const v = new Array<number>(limit).fill(0);
+      timestamps.set(username, v);
+      return v;
     }
+  })();
+  const index = indexes.get(username) ?? 0;
+  const current = timestamps_[index];
+  if (Date.now() - current <= timeframe) {
+    return timestamps_.filter((v) => v).sort((a, b) => a - b);
```

Comment:
Can you explain this?

Reply:
Wdym? Add a comment or just explain it to you here?

`timestamps_.filter((v) => v).sort((a, b) => a - b);` returns the timestamps that triggered the action so that they are displayed in the message.

Reply:
But why sort them? They're always sorted, aren't they?

Reply:
OK.

Reply:
No, not OK.

Reply:
They are not always sorted. What if it gets triggered in the first element?

Reply:
Then they're still two sorted lists that are disjoint

Reply:
I don't understand. Take this as an example, say the limit is 5:

10:11 10:12 10:13 10:14 10:15

It would then become:

10:16 10:17 10:13 10:14 10:15

...which is not sorted.

Reply:
And you can sort it by taking elements 2,3,4 and 0,1, there is no need to call `sort`. You even have the index stored at which the split happens. But hey, it's just 30 elements, so you could use bubble sort and it would be no difference. Just leave it.